### PR TITLE
feat: expose InitChain tx responses

### PR DESCRIPTION
**Pull request by n0izn0iz**
Save and allow to fetch genesis txs responses

### Before

```
❯ curl 'http://127.0.0.1:26657/block_results?height=0' -s | head -n 20
{
  "jsonrpc": "2.0",
  "id": "",
  "error": {
    "code": -32603,
    "message": "Internal error",
    "data": "height must be greater than 0"
  }
}
```

### After

```
❯ curl 'http://127.0.0.1:26657/block_results?height=0' -s | head -n 20
{
  "jsonrpc": "2.0",
  "id": "",
  "result": {
    "height": "0",
    "results": {
      "deliver_tx": [
        {
          "ResponseBase": {
            "Error": null,
            "Data": null,
            "Events": null,
            "Log": "msg:0,success:true,log:,events:[]",
            "Info": ""
          },
          "GasWanted": "50000",
          "GasUsed": "240261825"
        },
        {
          "ResponseBase": {
```

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: gno.land/pkg/gnoland/app.go [hunk 1]
```diff
@@ -180,11 +181,17 @@ func InitChainer(baseApp *sdk.BaseApp, acctKpr auth.AccountKeeperI, bankKpr bank
 			}
 
 			resHandler(ctx, tx, res)
+			txResponses = append(txResponses, abci.ResponseDeliverTx{
```
Comment (moul) [inline gno.land/pkg/gnoland/app.go:184]:
```suggestion
			txResponses[idx] = abci.ResponseDeliverTx{
```

**Comment by moul** [comment]
Is this common on other blockchains? I have concerns about the size of the response generated. I see advantages in having this new information, but I also think that it could be OK to just consider that "the genesis is there, and we don't need to know how."

**Comment by n0izn0iz** [comment]
I don't know if it's common in other blockchains but in order to index the portal loop we need to get the genesis transactions results otherwise the indexers can't replay after upgrade

**Comment by moul** [comment]
The Portal Loop isn't ideal for indexers. Indexers should focus on networks like test3 and test4. If we can enhance the Portal Loop to benefit other uses, we're open to it. Share how improving it could help beyond just the Portal Loop.

**Comment by leohhhn** [comment]
Hey @n0izn0iz, the Portal Loop is a bit specific when it comes to the way it works. It is not fully persistent will the way it handles state, and it can prove difficult to try and index it. 

We discussed this with Berty as well - we can either 1. add a way to fetch genesis transactions, 2. think of a different solution to index the Portal Loop. 

I'll bring this up on the engineer meeting today.

**Comment by n0izn0iz** [comment]
I also think it's important for other networks

What happens if you add a package in genesis that creates a grc20 token and want to index a "grc20.created" event, or "users.registered" events or dao membership history (some daos will probably be seeded in genesis, no?)

IMO there is many cases where it's desired to not have to extract this information in two ways, one for post-genesis-state via state inspection and one for regular txs via events

File: gno.land/pkg/gnoland/app.go (hunk 1 above)
Comment (n0izn0iz) [reply gno.land/pkg/gnoland/app.go:184]: I think this will panic since the cap is set, not the length

```
❯ cat main.go
package main

func main() {
	arr := make([]string, 0, 4)
	arr[0] = "foo"
}
❯ go run main.go
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
main.main()
	/Users/norman/Code/teritori-dapp/main.go:5 +0x24
exit status 2
```

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I'm a bit late to the party 🙏 

I like this approach, and I think it works fine, but I think we can do even better -- please check out my leftover comments regarding data saves

File: tm2/pkg/bft/consensus/replay.go [hunk 2]
```diff
@@ -306,6 +307,7 @@ func (h *Handshaker) ReplayBlocks(
 		if err != nil {
 			return nil, err
 		}
+		saveGenesisRes(h.stateDB, &res)
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go:310]: > not sure if this is possible, but worth checking out
block 0 is treated specially in complex procedures, I tried a bit this approach then fell back on the one in this PR, I'll try again though

**Comment by n0izn0iz** [comment]
sorry for the push --force, I push-forced again to restore history before squash

File: tm2/pkg/bft/consensus/replay.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go]: I improved to go more towards your way, the results are now queriable with rpc `/block_results?height=0` as shown in the new PR description

I still can't figure that part out

> genesis txs are indexable by hash from any client (or the node's RPC)

- in the genesis, txs are passed in a gno.land type, so should I move genesis transaction from GnoGenesisState to InitChainRequest? If you have another idea to get the genesis transactions hash pls tell me
- I don't see how we could get the transaction hashes to fetch in the first place, this would require us to also expose a genesis block and I feel this is way beyond the scope of this PR

File: tm2/pkg/bft/consensus/replay.go [hunk 3]
```diff
@@ -393,6 +395,15 @@
 		appBlockHeight, storeBlockHeight, stateBlockHeight))
 }
 
+// panics if failed to marshal the given genesis response
+func saveGenesisRes(db dbm.DB, res *abci.ResponseInitChain) {
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go]: this is removed in the new version

File: tm2/pkg/bft/consensus/replay_test.go [hunk 4]
```diff
@@ -1172,3 +1172,48 @@ func (ica *initChainApp) InitChain(req abci.RequestInitChain) abci.ResponseInitC
...
+
+type initTxsApp struct {
+	abci.BaseApplication
+}
+
+func (ica *initTxsApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChain {
+	return abci.ResponseInitChain{
+		TxResponses: make([]abci.ResponseDeliverTx, numInitResponses),
+	}
+}
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/consensus/replay_test.go:1219]: Please don't define these package-level, but within the test itself 🙏

File: tm2/pkg/bft/consensus/replay.go [hunk 5]
```diff
@@ -307,6 +307,13 @@ func (h *Handshaker) ReplayBlocks(
 			return nil, err
 		}
 
+		// Save the results by height
+		abciResponse := sm.NewABCIResponsesFromNum(len(res.TxResponses))
+		copy(abciResponse.DeliverTxs, res.TxResponses)
+		sm.SaveABCIResponses(h.stateDB, 0, abciResponse)
+
+		// NOTE: we don't save results by tx hash since the transactions are in the AppState opaque type
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/consensus/replay.go:315]: Here is a crazy idea: what do you think about saving the genesis block (`block 0`) to the DB?

I know we don't have all of the information for it, but I think having the majority of it fetch-able will allow for better integration with existing endpoints (specifically `tx`), and not require us to do the dance of (`min`) that we do for RPC endpoints in this PR. It just makes every client's life easier without having to "know" that `0` is unfetchable for _some_ endpoints (currently all)

**Review by zivkovicmilos — APPROVED** [review]
Much better direction, thank you for fixing this up 🫡 

I've left an important question about saving genesis blocks (full block) in the blockstore 🙏

File: tm2/pkg/bft/consensus/replay.go (hunk 5 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go:315]: I agree it would be better but I'm not sure how to do this without importing gno.land types into tm2

File: tm2/pkg/bft/consensus/replay_test.go (hunk 4 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 6c412202e2a8be14b3e06e367090640921b8b597

File: tm2/pkg/bft/consensus/replay_test.go [hunk 6]
```diff
@@ -1172,3 +1172,48 @@ func (ica *initChainApp) InitChain(req abci.RequestInitChain) abci.ResponseInitC
...
+
+	// check that the genesis transaction results are saved
+	res, err := sm.LoadABCIResponses(stateDB, 0)
+	if err != nil {
+		t.Fatalf("Failed to load genesis ABCI responses: %v", err)
+	}
+
+	if len(res.DeliverTxs) != numInitResponses {
+		t.Fatalf("Expected %d genesis tx responses, got %d", numInitResponses, len(res.DeliverTxs))
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 366d229c938080f6655e85ca8e3895c29fafcec0

File: tm2/pkg/bft/consensus/replay_test.go [hunk 7]
```diff
@@ -1172,3 +1172,48 @@ func (ica *initChainApp) InitChain(req abci.RequestInitChain) abci.ResponseInitC
...
+	defer proxyApp.Stop()
+	if err := handshaker.Handshake(proxyApp); err != nil {
+		t.Fatalf("Error on abci handshake: %v", err)
+	}
+
+	// check that the genesis transaction results are saved
+	res, err := sm.LoadABCIResponses(stateDB, 0)
+	if err != nil {
+		t.Fatalf("Failed to load genesis ABCI responses: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 366d229c938080f6655e85ca8e3895c29fafcec0

File: tm2/pkg/bft/consensus/replay_test.go [hunk 8]
```diff
@@ -1172,3 +1172,48 @@ func (ica *initChainApp) InitChain(req abci.RequestInitChain) abci.ResponseInitC
...
+	genDoc, _ := sm.MakeGenesisDocFromFile(genesisFile)
+	handshaker := NewHandshaker(stateDB, state, store, genDoc)
+	proxyApp := appconn.NewAppConns(clientCreator)
+	if err := proxyApp.Start(); err != nil {
+		t.Fatalf("Error starting proxy app connections: %v", err)
+	}
+	defer proxyApp.Stop()
+	if err := handshaker.Handshake(proxyApp); err != nil {
+		t.Fatalf("Error on abci handshake: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 366d229c938080f6655e85ca8e3895c29fafcec0

File: tm2/pkg/bft/consensus/replay_test.go [hunk 9]
```diff
@@ -1172,3 +1172,48 @@ func (ica *initChainApp) InitChain(req abci.RequestInitChain) abci.ResponseInitC
...
+	defer os.RemoveAll(config.RootDir)
+	stateDB, state, store := makeStateAndStore(config, genesisFile, "v0.0.0-test")
+
+	// now start the app using the handshake - it should sync
+	genDoc, _ := sm.MakeGenesisDocFromFile(genesisFile)
+	handshaker := NewHandshaker(stateDB, state, store, genDoc)
+	proxyApp := appconn.NewAppConns(clientCreator)
+	if err := proxyApp.Start(); err != nil {
+		t.Fatalf("Error starting proxy app connections: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 366d229c938080f6655e85ca8e3895c29fafcec0

File: tm2/pkg/bft/consensus/replay.go (hunk 5 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go:315]: There's also the whole block 0 vs block 1 ordeal

File: tm2/pkg/bft/abci/types/mock_application.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
+package abci
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/abci/types/mock_application.go:1]: Can we move this to an `internal` package if it's shared, or inside a `_test.go` file if not?

I'm usually not a huge fan of having mocks lay around like this in code, but this makes sense because I see how we can utilize it later on

File: tm2/pkg/bft/abci/types/types.go [hunk 11]
```diff
@@ -159,6 +159,7 @@ type ResponseInitChain struct {
 	ResponseBase
 	ConsensusParams *ConsensusParams
 	Validators      []ValidatorUpdate
+	TxResponses     []ResponseDeliverTx
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/abci/types/types.go:162]: This is valid 👍

File: tm2/pkg/bft/consensus/replay_test.go [hunk 12]
```diff
@@ -1144,13 +1152,11 @@ func TestHandshakeUpdatesValidators(t *testing.T) {
...
-	if err := proxyApp.Start(); err != nil {
-		t.Fatalf("Error starting proxy app connections: %v", err)
-	}
-	defer proxyApp.Stop()
-	if err := handshaker.Handshake(proxyApp); err != nil {
-		t.Fatalf("Error on abci handshake: %v", err)
-	}
+	err := proxyApp.Start()
+	require.NoError(t, err, "Error starting proxy app connections")
+	defer func() { err := proxyApp.Stop(); require.NoError(t, err) }()
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/consensus/replay_test.go:1157]: This error check can be inlined

File: tm2/pkg/bft/consensus/replay_test.go [hunk 13]
```diff
@@ -1144,13 +1152,11 @@ func TestHandshakeUpdatesValidators(t *testing.T) {
...
+	err = handshaker.Handshake(proxyApp)
+	require.NoError(t, err, "Error on abci handshake")
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/consensus/replay_test.go:1159]:
```suggestion
	require.NoError(t, handshaker.Handshake(proxyApp), "Error on abci handshake")
```

**Review by zivkovicmilos — APPROVED** [review]

**Comment by zivkovicmilos** [comment]
@moul 

We need you to sanity check us on this PR, the TL;DR is:
- @n0izn0iz exposed block results from the genesis block (`0`). This allows them to be queryable from the RPC
- @n0izn0iz also exposed the `DeliverTx` results on the InitChain response, which I think is valid. We swallowed this information in the past
- You still can't fetch these txs by hash (`tx` endpoint), since we can't save the tx index (you need full block information, which doesn't exist, we never "save" block 0 in the DB), but at least you can fetch the results

Does this sound good to you?

It unblocks us to actually index the genesis transactions from outside the node (albeit it does require a bit of custom logic for block `0`)

File: tm2/pkg/bft/abci/types/mock_application.go (hunk 10 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/abci/types/mock_application.go]: done in 404fcd59a71e7c55345608ccf2237fc24c1a0c10

I used `testing` as package name since the linter warns against importing stuff from a package named `internal`

Comment (n0izn0iz) [reply tm2/pkg/bft/abci/types/mock_application.go]: I could also move the `MockApplication` directly in the `consensus` package since it's only used here but I think it belongs more next the the `BaseApplication` definition

File: tm2/pkg/bft/consensus/replay_test.go (hunk 13 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 926e226f7944da16f18cf760413b2fdb1cc560ca

File: tm2/pkg/bft/consensus/replay_test.go (hunk 12 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 926e226f7944da16f18cf760413b2fdb1cc560ca

File: tm2/pkg/bft/consensus/replay.go (hunk 5 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay.go:315]: For lurkers:

Block 1 is using the state derived in InitChain as base (it's a special case in BeginBlock) but is pulling txs from the mempool, so genesis transasction seem to belong more to some "Block 0"

File: tm2/pkg/bft/state/store.go [hunk 14]
```diff
@@ -131,8 +132,13 @@ type ABCIResponses struct {
 
 // NewABCIResponses returns a new ABCIResponses
 func NewABCIResponses(block *types.Block) *ABCIResponses {
-	resDeliverTxs := make([]abci.ResponseDeliverTx, block.NumTxs)
-	if block.NumTxs == 0 {
+	return NewABCIResponsesFromNum(block.NumTxs)
+}
+
+// NewABCIResponsesFromNum returns a new ABCIResponses with a set number of txs
+func NewABCIResponsesFromNum[N constraints.Integer](numTxs N) *ABCIResponses {
```
Comment (moul) [inline tm2/pkg/bft/state/store.go:139]: couldn't we just use `int`, here?

**Comment by moul** [comment]
What about a server starting with a quick sync snapshot, beginning at block 1M?

**Comment by moul** [comment]
The Genesis file is approximately 700MB on test3.

File: tm2/pkg/bft/state/store.go (hunk 14 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/state/store.go:139]: there is two places where this is used, one is int64, one is int
we could use a cast a call site but I thought this would be better, tell me if you want me to change this

**Comment by n0izn0iz** [comment]
I don't fully understand your concerns, the full genesis goes in the ABCI protocol and it can be fetched from the api currently. I don't understand why we couldn't fetch the results also

File: tm2/pkg/bft/state/store.go (hunk 14 above)
Comment (moul) [reply tm2/pkg/bft/state/store.go:139]: please, use a cast call and keep things simple.

**Comment by moul** [comment]
> > What about a server starting with a quick sync snapshot, beginning at block 1M?
> I don't understand the question

What happens when someone starts at block 1M instead of block 0? Will your virtual block 1M differ from other nodes that started from an earlier genesis? Could you add a test to expose this case for a chain starting with a sync snapshot?

> > The Genesis file is approximately 700MB on test3.
> The full genesis goes in the ABCI protocol and it can be fetched from the api currently. I don't understand why we couldn't return and expose the results also

I think I don't fully understand the purpose of this PR. Could someone please explain the different elements (genesis, virtual block 0, and upcoming blocks) and how they are affected by this PR, perhaps with a visual schema?

File: tm2/pkg/bft/consensus/replay_test.go [hunk 15]
```diff
@@ -1161,14 +1166,34 @@ func TestHandshakeUpdatesValidators(t *testing.T) {
...
+		InitChainFn: func(req abci.RequestInitChain) abci.ResponseInitChain {
+			return abci.ResponseInitChain{
+				TxResponses: make([]abci.ResponseDeliverTx, numInitResponses),
+			}
+		},
 	}
+	clientCreator := proxy.NewLocalClientCreator(app)
+
+	config, genesisFile := ResetConfig("handshake_test_")
+	defer os.RemoveAll(config.RootDir)
```
Comment (ajnavarro) [inline tm2/pkg/bft/consensus/replay_test.go:1184]:
```suggestion
	t.Cleanup(func() {
	    os.RemoveAll(config.RootDir)
	})
```

File: tm2/pkg/bft/consensus/replay_test.go [hunk 16]
```diff
@@ -1144,13 +1153,9 @@ func TestHandshakeUpdatesValidators(t *testing.T) {
...
 	proxyApp := appconn.NewAppConns(clientCreator)
-	if err := proxyApp.Start(); err != nil {
-		t.Fatalf("Error starting proxy app connections: %v", err)
-	}
-	defer proxyApp.Stop()
-	if err := handshaker.Handshake(proxyApp); err != nil {
-		t.Fatalf("Error on abci handshake: %v", err)
-	}
+	require.NoError(t, proxyApp.Start(), "Error starting proxy app connections")
+	defer func() { require.NoError(t, proxyApp.Stop()) }()
```
Comment (ajnavarro) [inline tm2/pkg/bft/consensus/replay_test.go:1157]:
```suggestion
       t.Cleanup(func() {
	    require.NoError(t, proxyApp.Stop())
	})
```

File: tm2/pkg/bft/consensus/replay_test.go [hunk 17]
```diff
@@ -1161,14 +1166,34 @@ func TestHandshakeUpdatesValidators(t *testing.T) {
...
+	config, genesisFile := ResetConfig("handshake_test_")
+	defer os.RemoveAll(config.RootDir)
+	stateDB, state, store := makeStateAndStore(config, genesisFile, "v0.0.0-test")
+
+	// now start the app using the handshake - it should sync
+	genDoc, _ := sm.MakeGenesisDocFromFile(genesisFile)
+	handshaker := NewHandshaker(stateDB, state, store, genDoc)
+	proxyApp := appconn.NewAppConns(clientCreator)
+	require.NoError(t, proxyApp.Start(), "Error starting proxy app connections")
+	defer func() { require.NoError(t, proxyApp.Stop()) }()
```
Comment (ajnavarro) [inline tm2/pkg/bft/consensus/replay_test.go:1192]:
```suggestion
        t.Cleanup(func(){
            require.NoError(t, proxyApp.Stop())
        })
```

File: tm2/pkg/bft/rpc/core/blocks.go [hunk 18]
```diff
@@ -417,11 +417,11 @@ func BlockResults(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlockR
...
 }
 
-func getHeight(currentHeight int64, heightPtr *int64) (int64, error) {
+func getHeight(currentHeight int64, heightPtr *int64, min int64) (int64, error) {
 	if heightPtr != nil {
 		height := *heightPtr
-		if height <= 0 {
-			return 0, fmt.Errorf("height must be greater than 0")
+		if height < min {
+			return 0, fmt.Errorf("height must be greater than or equal to %d", min)
```
Comment (ajnavarro) [inline tm2/pkg/bft/rpc/core/blocks.go:424]: the error and the if expression are not the same:

```suggestion
		if height <= min {
			return 0, fmt.Errorf("height must be greater than or equal to %d", min)
```

Comment (ajnavarro) [reply tm2/pkg/bft/rpc/core/blocks.go:424]: shall we add a test for this corner case?

File: tm2/pkg/bft/rpc/core/blocks.go [hunk 19]
```diff
@@ -417,11 +417,11 @@ func BlockResults(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlockR
...
 		}
 		if height > currentHeight {
```
Comment (ajnavarro) [inline tm2/pkg/bft/rpc/core/blocks.go:426]: The same is happening here...

**Review by ajnavarro — CHANGES_REQUESTED** [review]

File: tm2/pkg/bft/rpc/core/blocks.go (hunk 18 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/rpc/core/blocks.go:424]: mmh, the check and error are matching, the error message must be the inverse of the if expression

if `height` is less than `min`, then we stop and return an error saying that `height` must be greater than or equal to `min`

it could be rewritten as `if !(height >= min) {` if you think it's clearer

I will add a test

File: tm2/pkg/bft/consensus/replay_test.go (hunk 16 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 87fe7d8b6f2f145dc4ef2cecb6577d15ab40d3ba

File: tm2/pkg/bft/consensus/replay_test.go (hunk 17 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/consensus/replay_test.go]: done in 87fe7d8b6f2f145dc4ef2cecb6577d15ab40d3ba

File: tm2/pkg/bft/rpc/core/blocks.go (hunk 18 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/rpc/core/blocks.go:424]: test added in bad682a272123a943fd2163cb501dbadc837500b

**Comment by n0izn0iz** [comment]
@moul

> What happens when someone starts at block 1M instead of block 0? Will your virtual block 1M differ from other nodes that started from an earlier genesis? Could you add a test to expose this case for a chain starting with a sync snapshot?

How could I start at block 1M? I don't see a `StartHeight` field in any init types (neither GenesisDoc, nor InitChainRequest)

> I think I don't fully understand the purpose of this PR. Could someone please explain the different elements (genesis, virtual block 0, and upcoming blocks) and how they are affected by this PR, perhaps with a visual schema?

The way a tendermint chain is initialized is through the `InitChain` ABCI endpoint

The in-memory `GenesisDoc` is passed to this call, in the case of gno (and most tendermint chains), this GenesisDoc contains transactions that are executed during the `InitChain` call

Currently, the result of these transactions is lost, this is very annoying for indexing as, among other things, you loose events from those transactions.

This PR returns the result of these genesis transactions in the `InitChain` response.

These results are stored as belonging to `block #0` because the transactions in `block #1` come from the mempool and are executed after the `InitChain` routine

**Review by ajnavarro — APPROVED** [review]

File: tm2/pkg/bft/state/store.go [hunk 20]
```diff
@@ -175,7 +180,7 @@ func LoadABCIResponses(db dbm.DB, height int64) (*ABCIResponses, error) {
 // SaveABCIResponses persists the ABCIResponses to the database.
 // This is useful in case we crash after app.Commit and before s.Save().
 // Responses are indexed by height so they can also be loaded later to produce Merkle proofs.
-func saveABCIResponses(db dbm.DB, height int64, abciResponses *ABCIResponses) {
+func SaveABCIResponses(db dbm.DB, height int64, abciResponses *ABCIResponses) {
```
Comment (thehowl) [inline tm2/pkg/bft/state/store.go:183]:
```suggestion
// NOTE: this should only be used internally by the bft package and subpackages.
func SaveABCIResponses(db dbm.DB, height int64, abciResponses *ABCIResponses) {
```

Comment (thehowl) [reply tm2/pkg/bft/state/store.go:183]: I'd suggest moving it in an internal/ package, but it looks complex to do, so let's roll with this for now.

**Review by thehowl — APPROVED** [review]

File: tm2/pkg/bft/rpc/core/blocks.go [hunk 21]
```diff
@@ -400,7 +400,7 @@ func Commit(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultCommit, erro
 // ```
 func BlockResults(ctx *rpctypes.Context, heightPtr *int64) (*ctypes.ResultBlockResults, error) {
 	storeHeight := blockStore.Height()
-	height, err := getHeight(storeHeight, heightPtr)
+	height, err := getHeight(storeHeight, heightPtr, 0)
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/rpc/core/blocks.go]: done in 110da7d8c9e1cebceb069d73077f02072b07040f

File: tm2/pkg/bft/abci/types/testing/mock_application.go
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply tm2/pkg/bft/abci/types/testing/mock_application.go:1]: done in 01dc6f8365b4f1a7ac673b73dc286ba219a6b001

File: tm2/pkg/bft/state/store.go (hunk 20 above)
Comment (n0izn0iz) [reply tm2/pkg/bft/state/store.go:184]: applied in [e649de5](https://github.com/gnolang/gno/pull/1941/commits/e649de519ea1fd7ec60399c7ab4df0b85c7e8cde)

**Review by thehowl — APPROVED** [review]

**Comment by n0izn0iz** [comment]
❤️
